File: stack/stack.rs
// use std::io;

#[derive(Debug)]
struct Stack<T> {
    head: Option<T>,
    tail: Option<Box<Stack<T>>>,
}

impl<T: Copy> Stack<T> {
    fn new() -> Stack<T> {
        Stack {
            head: None,
            tail: None,
        }
    }

    fn push(&mut self, data: T) {
        if self.head.is_none() {
            self.head = Some(data);
            return;
        }
        match self.tail {
            Some(ref mut current_data) => current_data.push(data),
            None => {
                self.tail = Some(Box::new(Stack{
                    head: Some(data),
                    tail: None,
                }));  
            }
        }
    }

    fn pop(&mut self) -> Option<T> {
        let temp;
        match self.tail {
            Some(ref mut stack) => {
                if stack.tail.is_some() {
                    return stack.pop();
                } else {
                    temp = stack.head;
                }
            },
            None => {
                temp = self.head;
                self.head = None;
                return temp;
            },
        }
        self.tail = None;
        return temp;
    }
}

fn main() {
    let mut a = Stack::new();
    let b = vec![1,2,3,4,3,4,7];
    for i in b {
        a.push(i);
    }
    println!("{:?}", a.pop());
    println!("{:?}", a.pop());
    println!("{:?}", a.pop());
    println!("{:?}", a.pop());
    println!("{:?}", a.pop());
    // println!("{:?}", a.pop());
    // println!("{:?}", a.pop());
    // println!("{:?}", a.pop());
    println!("{:?}", a);
}

File: stack/stack.c
#include <stdio.h>

typedef struct Stack {
    char *head;
    struct Stack *tail;
} Stack;

Stack newStack() {
    static Stack temp;
    temp.head = NULL;
    temp.tail = NULL;
    return temp;
}

void push(Stack *stack, char *data) {
    if (stack->head == NULL) {
        stack->head = data;
        return;
    }
    if (stack->tail == NULL) {
        Stack ns = newStack();
        ns.head = data;
        stack->tail = &ns;
    } else {
        push(stack->tail, data);
    }
}

char *pop(Stack *stack) {
    char *temp;
    if (stack->tail == NULL) {
        temp = stack->head;
        stack->head = NULL;
        return temp;
    } else {
        if (stack->tail->tail == NULL) {
            temp = stack->tail->head;
            stack->tail = NULL;
            return temp;
        } else {
            return pop(stack->tail);
        }
    }
}

int main(void) {

    Stack a = newStack();
    push(&a, "12");
    push(&a, "232123213");
    // char *ptr = a.head;
    printf("%s\n", pop(&a));
    printf("%s\n", pop(&a));

    return 0;
}